### DISCUSSION TITLE
How to work with cookies in insomnia.sendRequest (Pre-Request and Post-Request scripts)?

### DISCUSSION THREAD

**Discussion by andreygrigoriev** [Kong/insomnia#7791] [Help] [answered]
Hi. First of all thank you for implementing Pre-Request and Post-Request scripts. This is a feature we waited for a long time. Several questions regarding Pre-Request and Post-Request scripts:

1. Are cookies stored somewhere between several `insomnia.sendRequest` executions?
2. Is it possible to add a custom cookie to the `insomnia.sendRequest` execution?
3. Is it possible to show the complete log of the request-response (headers, cookies, everything) for `insomnia.sendRequest` execution?
4. Is there any documentation on `insomnia.sendRequest`?

**Comment by subnetmarco** [comment]
@ihexxa

**Comment by ihexxa** [comment, marked as answer]
Hi @andreygrigoriev , thanks for trying scripts.

1. Nope, currently cookies are persisted only after script execution completion, so you might maintain cookie values manually.
2. Yes, although it is a bit complicated, will try to simplify it:
```
const { Cookie } = require('insomnia-collection');
const cookies = [
	new Cookie({key: 'myCookie1', value: 'abc'}),
	new Cookie({key: 'myCookie2', value: 'def'}),
];

const rawReq = {
    url: 'httpbin.org/anything',
    method: 'POST',
    header: {
	'Cookie': Cookie.unparse(cookies),
        'Content-Type': 'text/plain',
    },
    body: {
        mode: 'raw',
        raw: 'rawContent',
    },
};

await insomnia.sendRequest(
	rawReq,
	(error, response) => {
		if (error) {
			console.log(error);
		} else {
			console.log(response);
		}
	}
);
```

3. IIUC, you could try to define one request manually [doc](https://docs.insomnia.rest/insomnia/pre-request-script#send-a-request-advanced-mode), and `console.log` it. And you can also see how to print the response in the example.

4. Yes, you can find some detailed examples in the above doc.

**Reply by andreygrigoriev** [reply]
Thanks!
